File: python/balance_a_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        nodes = []
        #add the nodes to a list using in-order traversal
        def visit(node: TreeNode) -> None:
            if node is None:
                return
            visit(node.left)
            nodes.append(node)
            visit(node.right)
        visit(root)
        root_index = len(nodes) // 2
        new_root = nodes[root_index]
        #each subproblem takes the root index and the left and right bounds of the nodes array
        #it assigns the left child to the node halfway between the left bound and the node index and the
        #right child to the node halfway between the node index and the right bound
        def add_descendants(node_index: int, l: int, r: int) -> None:
            left_l, left_r = l, node_index - 1
            right_l, right_r = node_index + 1, r
            if left_l <= left_r:
                left_child_index = left_l + (left_r - left_l) // 2
                nodes[node_index].left = nodes[left_child_index]
                add_descendants(left_child_index, left_l, left_r)
            else:
                nodes[node_index].left = None
            if right_l <= right_r:
                right_child_index = right_l + (right_r - right_l) // 2
                nodes[node_index].right = nodes[right_child_index]
                add_descendants(right_child_index, right_l, right_r)
            else:
                nodes[node_index].right = None
        add_descendants(root_index, 0, len(nodes) - 1)
        return new_root

File: java/balanceABinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //will store nodes in order
    List<TreeNode> nodes = new ArrayList();
    
    public TreeNode balanceBST(TreeNode root) {
        //traverse tree in order and store the nodes in the right order in a list
        visit(root);
        //each subproblem of addDescendants takes a left and right bound and the index of the parent element (which will be in the middle)
        int leftBound = 0, middle = (nodes.size() - 1) / 2, rightBound = nodes.size() - 1;
        addDescendants(leftBound, middle, rightBound);
        return nodes.get(middle);
    }
    
    private void addDescendants(int leftBound, int middle, int rightBound) {
        //if there is at least one element to the left of the parent element, add it as the left child
        int leftRight = middle - 1, rightLeft = middle + 1;
        if (leftRight >= leftBound) {
            int leftMiddle = leftBound + (leftRight - leftBound) / 2;
            nodes.get(middle).left = nodes.get(leftMiddle);
            addDescendants(leftBound, leftMiddle, leftRight);
        } else {
            nodes.get(middle).left = null;
        }
        //if there is at least one element to the right of the parent element, add it as the right child
        if (rightLeft <= rightBound) {
            int rightMiddle = rightLeft + (rightBound - rightLeft) / 2;
            nodes.get(middle).right = nodes.get(rightMiddle);
            addDescendants(rightLeft, rightMiddle, rightBound);
        } else {
            nodes.get(middle).right = null;
        }
    }
    
    private void visit(TreeNode node) {
        //uses in-order traversal to add nodes to nodes list
        if (node != null) {
            visit(node.left);
            nodes.add(node);
            visit(node.right);
        }
    }
}